From 612b4abeb4a2d46b4b665f988f7c1db4c34b3fd6 Mon Sep 17 00:00:00 2001
From: Jarno Rajahalme <jarno@isovalent.com>
Date: Mon, 5 May 2025 11:15:52 +1000
Subject: [PATCH 1/6] network: Add callback for upstream authorization

Add new ReadFilterCallbacks addUpstreamCallback() and
iterateUpstreamCallbacks(). Network filters can add callbacks using
addUpstreamCallback(), which will then get called after an upstream
host has been selected, but before the upstream connection is
established.  If any of the callbacks returns 'false', the connection
is not established. For HTTP the router will issue a 403 local
response.

iterateUpstreamCallbacks() is also added to
StreamDecoderFilterCallbacks so that the HTTP router filter can invoke
the added callbacks before a new connection is established.

These additions allow network read filters to perform network level
policy enforcement based on the selected upstream host.

Callbacks can safely refer to memory held by the filter instance
adding the callback, as the calls to the callbacks are only ever be
done from the tcp_proxy or router filter in the same filter chain.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
---
 envoy/http/filter.h                           |  8 ++++++
 envoy/network/filter.h                        | 28 +++++++++++++++++++
 envoy/tcp/upstream.h                          |  5 ++++
 source/common/http/async_client_impl.h        |  5 ++++
 source/common/http/conn_manager_impl.h        |  6 ++++
 source/common/http/filter_manager.cc          |  6 ++++
 source/common/http/filter_manager.h           |  8 ++++++
 source/common/network/filter_manager_impl.h   | 21 ++++++++++++++
 source/common/router/router.cc                |  8 ++++++
 source/common/router/upstream_request.h       |  5 ++++
 source/common/tcp_proxy/tcp_proxy.cc          |  7 +++++
 source/common/tcp_proxy/tcp_proxy.h           |  5 ++++
 source/common/tcp_proxy/upstream.cc           |  8 ++++++
 source/common/tcp_proxy/upstream.h            |  2 ++
 .../default_api_listener/api_listener_impl.h  |  3 ++
 15 files changed, 125 insertions(+)

diff --git a/envoy/http/filter.h b/envoy/http/filter.h
index 91d59de40b..d94b89f0e7 100644
--- a/envoy/http/filter.h
+++ b/envoy/http/filter.h
@@ -829,6 +829,14 @@ public:
   virtual absl::optional<Upstream::LoadBalancerContext::OverrideHost>
   upstreamOverrideHost() const PURE;

+  /**
+   * Invokes all the added network level callbacks before establishing a connection to the
+   * selected upstream host.
+   * Returns 'false' if any of the callbacks rejects the connection, 'true' otherwise.
+   */
+  virtual bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr,
+                                        StreamInfo::StreamInfo&) PURE;
+
   /**
    * @return true if the filter should shed load based on the system pressure, typically memory.
    */
diff --git a/envoy/network/filter.h b/envoy/network/filter.h
index 8684036d18..9f3f4a42d9 100644
--- a/envoy/network/filter.h
+++ b/envoy/network/filter.h
@@ -147,6 +147,22 @@ public:

 using WriteFilterSharedPtr = std::shared_ptr<WriteFilter>;

+/**
+ * UpstreamCallback can be used to reject upstream host selection made by the TCP proxy filter.
+ * This callback is passed the Upstream::HostDescriptionConstSharedPtr, and StreamInfo.
+ *
+ * The callback is called just after the upstream host has been picked, but before a connection is
+ * established. Here the callback can reject the selected upstream host and cause the be dropped.
+
+ * UpstreamCallback may not be called if the connection is dropped for another reason, such as
+ * no route, cluster is not found, etc.
+ *
+ * Returning 'true' allows the connection to be established. Returning 'false' prevents the
+ * connection to the selected host from being established.
+ */
+using UpstreamCallback = std::function<bool(Upstream::HostDescriptionConstSharedPtr,
+                                            StreamInfo::StreamInfo&)>;
+
 /**
  * Callbacks used by individual read filter instances to communicate with the filter manager.
  */
@@ -206,6 +222,18 @@ public:
    */
   virtual bool startUpstreamSecureTransport() PURE;

+  /*
+   * Adds the given callback to be executed later via iterateUpstreamCallbacks().
+   */
+  virtual void addUpstreamCallback(const UpstreamCallback& cb) PURE;
+
+  /**
+   * Invokes all the added callbacks before connecting to the selected upstream host.
+   * Returns 'false' if any of the callbacks rejects the connection, 'true' otherwise.
+   */
+  virtual bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr,
+                                        StreamInfo::StreamInfo&) PURE;
+
   /**
    * Control the filter close status for read filters.
    *
diff --git a/envoy/tcp/upstream.h b/envoy/tcp/upstream.h
index b201d2e153..9aece0fc6c 100644
--- a/envoy/tcp/upstream.h
+++ b/envoy/tcp/upstream.h
@@ -72,6 +72,11 @@ public:
    * @param callbacks callbacks to communicate stream failure or creation on.
    */
   virtual void newStream(GenericConnectionPoolCallbacks& callbacks) PURE;
+
+  /**
+   * @return Upstream::HostDescriptionConstSharedPtr the host for which connections are pooled.
+   */
+  virtual Upstream::HostDescriptionConstSharedPtr host() const PURE;
 };

 // An API for the UpstreamRequest to get callbacks from either an HTTP or TCP
diff --git a/source/common/http/async_client_impl.h b/source/common/http/async_client_impl.h
index 9e0bc1248d..af2da3f9bd 100644
--- a/source/common/http/async_client_impl.h
+++ b/source/common/http/async_client_impl.h
@@ -264,6 +264,11 @@ private:
   ResponseHeaderMapOptRef responseHeaders() override { return {}; }
   ResponseTrailerMapOptRef responseTrailers() override { return {}; }

+  bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr,
+                            StreamInfo::StreamInfo&) override {
+    return true;
+  }
+
   // ScopeTrackedObject
   void dumpState(std::ostream& os, int indent_level) const override {
     const char* spaces = spacesForLevel(indent_level);
diff --git a/source/common/http/conn_manager_impl.h b/source/common/http/conn_manager_impl.h
index 812530b06c..4178eee81b 100644
--- a/source/common/http/conn_manager_impl.h
+++ b/source/common/http/conn_manager_impl.h
@@ -324,6 +324,12 @@ private:
     }

     absl::optional<Router::ConfigConstSharedPtr> routeConfig();
+
+    bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr host,
+          StreamInfo::StreamInfo& stream_info) const override {
+      return connection_manager_.read_callbacks_->iterateUpstreamCallbacks(host, stream_info);
+    }
+
     void traceRequest();

     // Updates the snapped_route_config_ (by reselecting scoped route configuration), if a scope is
diff --git a/source/common/http/filter_manager.cc b/source/common/http/filter_manager.cc
index 2db8bec5d0..b35f3ad707 100644
--- a/source/common/http/filter_manager.cc
+++ b/source/common/http/filter_manager.cc
@@ -1982,5 +1982,11 @@ ActiveStreamDecoderFilter::upstreamOverrideHost() const {
       parent_.upstream_override_host_.second};
 }

+bool ActiveStreamDecoderFilter::iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr host,
+                                                         StreamInfo::StreamInfo& stream_info) {
+  return parent_.filter_manager_callbacks_.iterateUpstreamCallbacks(host, stream_info);
+
+}
+
 } // namespace Http
 } // namespace Envoy
diff --git a/source/common/http/filter_manager.h b/source/common/http/filter_manager.h
index ee5391bf06..40f300eb73 100644
--- a/source/common/http/filter_manager.h
+++ b/source/common/http/filter_manager.h
@@ -305,6 +305,8 @@ struct ActiveStreamDecoderFilter : public ActiveStreamFilterBase,
   void setUpstreamOverrideHost(Upstream::LoadBalancerContext::OverrideHost) override;
   absl::optional<Upstream::LoadBalancerContext::OverrideHost> upstreamOverrideHost() const override;
   bool shouldLoadShed() const override;
+  bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr host,
+                              StreamInfo::StreamInfo& stream_info) override;
   void sendGoAwayAndClose() override;

   // Each decoder filter instance checks if the request passed to the filter is gRPC
@@ -591,6 +593,12 @@ public:
    * This is used for HTTP/1.1 codec.
    */
   virtual bool isHalfCloseEnabled() PURE;
+
+  /*
+   * Returns whether connection to the selected upstream host is allowed.
+   */
+  virtual bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr,
+                                        StreamInfo::StreamInfo&) const PURE;
 };

 /**
diff --git a/source/common/network/filter_manager_impl.h b/source/common/network/filter_manager_impl.h
index 6453048610..d4132a33ca 100644
--- a/source/common/network/filter_manager_impl.h
+++ b/source/common/network/filter_manager_impl.h
@@ -156,6 +156,13 @@ private:
       parent_.host_description_ = host;
     }
     bool startUpstreamSecureTransport() override { return parent_.startUpstreamSecureTransport(); }
+    void addUpstreamCallback(const UpstreamCallback& cb) override {
+      parent_.addUpstreamCallback(cb);
+    }
+    bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr host,
+                                  StreamInfo::StreamInfo& stream_info) override {
+      return parent_.iterateUpstreamCallbacks(host, stream_info);
+    }

     FilterManagerImpl& parent_;
     ReadFilterSharedPtr filter_;
@@ -190,6 +197,20 @@ private:
   FilterStatus onWrite(ActiveWriteFilter* filter, WriteBufferSource& buffer_source);
   void onResumeWriting(ActiveWriteFilter* filter, WriteBufferSource& buffer_source);

+  void addUpstreamCallback(const UpstreamCallback& cb) {
+    decoder_filter_upstream_cbs_.emplace_back(cb);
+  }
+
+  bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr host,
+        StreamInfo::StreamInfo& stream_info) {
+    bool accept = true;
+    for (const auto& cb : decoder_filter_upstream_cbs_) {
+      accept = accept && cb(host, stream_info);
+    }
+    return accept;
+  }
+
+  std::vector<UpstreamCallback> decoder_filter_upstream_cbs_{};
   FilterManagerConnection& connection_;
   const ConnectionSocket& socket_;
   Upstream::HostDescriptionConstSharedPtr host_description_;
diff --git a/source/common/router/router.cc b/source/common/router/router.cc
index a01ff69ee1..f5f2807f06 100644
--- a/source/common/router/router.cc
+++ b/source/common/router/router.cc
@@ -748,6 +748,14 @@ bool Filter::continueDecodeHeaders(Upstream::ThreadLocalCluster* cluster,
     return false;
   }

+  bool accepted = callbacks_->iterateUpstreamCallbacks(host, callbacks_->streamInfo());
+  if (!accepted) {
+    callbacks_->streamInfo().setResponseFlag(StreamInfo::CoreResponseFlag::UnauthorizedExternalService);
+    callbacks_->sendLocalReply(Http::Code::Forbidden, "Access denied\r\n",
+                               nullptr, absl::nullopt, absl::string_view());
+    return false;
+  }
+
   hedging_params_ = FilterUtility::finalHedgingParams(*route_entry_, headers);

   timeout_ = FilterUtility::finalTimeout(*route_entry_, headers, !config_->suppress_envoy_headers_,
diff --git a/source/common/router/upstream_request.h b/source/common/router/upstream_request.h
index 755e4e2fad..8842169be2 100644
--- a/source/common/router/upstream_request.h
+++ b/source/common/router/upstream_request.h
@@ -352,6 +352,11 @@ public:
   }
   OptRef<UpstreamStreamFilterCallbacks> upstreamCallbacks() override { return {*this}; }

+  bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr,
+        StreamInfo::StreamInfo&) const override {
+    return true;
+  }
+
   // Http::UpstreamStreamFilterCallbacks
   StreamInfo::StreamInfo& upstreamStreamInfo() override { return upstream_request_.streamInfo(); }
   OptRef<GenericUpstream> upstream() override {
diff --git a/source/common/tcp_proxy/tcp_proxy.cc b/source/common/tcp_proxy/tcp_proxy.cc
index 656ba58a4f..e972b8a8bd 100644
--- a/source/common/tcp_proxy/tcp_proxy.cc
+++ b/source/common/tcp_proxy/tcp_proxy.cc
@@ -649,6 +649,13 @@ bool Filter::maybeTunnel(Upstream::ThreadLocalCluster& cluster) {
         upstream_decoder_filter_callbacks_, getStreamInfo());
   }
   if (generic_conn_pool_) {
+    bool accepted = read_callbacks_->iterateUpstreamCallbacks(generic_conn_pool_->host(), getStreamInfo());
+    if (!accepted) {
+      getStreamInfo().setResponseFlag(StreamInfo::CoreResponseFlag::UnauthorizedExternalService);
+      onInitFailure(UpstreamFailureReason::UnauthorizedExternalService);
+      return true;
+    }
+
     connecting_ = true;
     connect_attempts_++;
     getStreamInfo().setAttemptCount(connect_attempts_);
diff --git a/source/common/tcp_proxy/tcp_proxy.h b/source/common/tcp_proxy/tcp_proxy.h
index a4cd362afd..c78a89c964 100644
--- a/source/common/tcp_proxy/tcp_proxy.h
+++ b/source/common/tcp_proxy/tcp_proxy.h
@@ -564,6 +564,10 @@ public:
       return absl::nullopt;
     }
     bool shouldLoadShed() const override { return false; }
+    bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr host,
+                                                     StreamInfo::StreamInfo& stream_info) override {
+      return parent_->upstream_decoder_filter_callbacks_.iterateUpstreamCallbacks(host, stream_info);
+    }
     void restoreContextOnContinue(ScopeTrackedObjectStack& tracked_object_stack) override {
       tracked_object_stack.add(*this);
     }
@@ -607,6 +611,7 @@ protected:
     NoHealthyUpstream,
     ResourceLimitExceeded,
     NoRoute,
+    UnauthorizedExternalService,
   };

   // Callbacks for different error and success states during connection establishment
diff --git a/source/common/tcp_proxy/upstream.cc b/source/common/tcp_proxy/upstream.cc
index 6c9e93c948..857f654dc0 100644
--- a/source/common/tcp_proxy/upstream.cc
+++ b/source/common/tcp_proxy/upstream.cc
@@ -247,6 +247,10 @@ void TcpConnPool::newStream(GenericConnectionPoolCallbacks& callbacks) {
   }
 }

+Upstream::HostDescriptionConstSharedPtr TcpConnPool::host() const {
+  return conn_pool_data_.value().host();
+}
+
 void TcpConnPool::onPoolFailure(ConnectionPool::PoolFailureReason reason,
                                 absl::string_view failure_reason,
                                 Upstream::HostDescriptionConstSharedPtr host) {
@@ -353,6 +357,10 @@ void HttpConnPool::newStream(GenericConnectionPoolCallbacks& callbacks) {
   }
 }

+Upstream::HostDescriptionConstSharedPtr HttpConnPool::host() const {
+  return conn_pool_data_.value().host();
+}
+
 void HttpConnPool::onPoolFailure(ConnectionPool::PoolFailureReason reason,
                                  absl::string_view failure_reason,
                                  Upstream::HostDescriptionConstSharedPtr host) {
diff --git a/source/common/tcp_proxy/upstream.h b/source/common/tcp_proxy/upstream.h
index 661797dc65..781a7549a5 100644
--- a/source/common/tcp_proxy/upstream.h
+++ b/source/common/tcp_proxy/upstream.h
@@ -40,6 +40,7 @@ public:

   // GenericConnPool
   void newStream(GenericConnectionPoolCallbacks& callbacks) override;
+  Upstream::HostDescriptionConstSharedPtr host() const override;

   // Tcp::ConnectionPool::Callbacks
   void onPoolFailure(ConnectionPool::PoolFailureReason reason,
@@ -97,6 +98,7 @@ public:

   // GenericConnPool
   void newStream(GenericConnectionPoolCallbacks& callbacks) override;
+  Upstream::HostDescriptionConstSharedPtr host() const override;

   // Http::ConnectionPool::Callbacks,
   void onPoolFailure(ConnectionPool::PoolFailureReason reason,
diff --git a/source/extensions/api_listeners/default_api_listener/api_listener_impl.h b/source/extensions/api_listeners/default_api_listener/api_listener_impl.h
index 01ec2f3015..bd1b3cfbe6 100644
--- a/source/extensions/api_listeners/default_api_listener/api_listener_impl.h
+++ b/source/extensions/api_listeners/default_api_listener/api_listener_impl.h
@@ -81,6 +81,9 @@ protected:
     }
     Network::Connection& connection() override { return connection_; }
     const Network::ConnectionSocket& socket() override { PANIC("not implemented"); }
+    void addUpstreamCallback(const Network::UpstreamCallback&) override {}
+    bool iterateUpstreamCallbacks(Upstream::HostDescriptionConstSharedPtr,
+				  StreamInfo::StreamInfo&) override { return true; }

     // Synthetic class that acts as a stub for the connection backing the
     // Network::ReadFilterCallbacks.
--
2.49.0

